annet/mesh/basemodel.py

```diff
@@ -0,0 +1,155 @@
+from abc import ABC
+from copy import copy
+from enum import Enum
+from typing import TypeVar, Any, Annotated, get_origin, get_type_hints, get_args, Callable
+
+
+class MergeForbiddenError(Exception):
+    pass
+
+
+class Special(Enum):
+    NOT_SET = "<NOT SET>"
+
+
+T = TypeVar('T')
+
+
+class Merger(ABC):
+    def __call__(self, name: str, x: T | Special, y: T | Special) -> T:
+        if x is Special.NOT_SET:
+            return y
+        if y is Special.NOT_SET:
+            return x
+        return self._merge(name, x, y)
+
+    def _merge(self, name: str, x: T, y: T) -> T:
+        raise NotImplementedError
+
+
+class UseFirst(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        return x
+
+
+class UseLast(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        return y
+
+
+class Forbid(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        raise MergeForbiddenError(f"Override is forbidden for field {name}")
+
+
+class ForbidChange(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        if x == y:
+            return x
+        raise MergeForbiddenError(
+            f"Override with different value is forbidden for field {name}:\n"
+            f"Old: {x}\n"
+            f"New: {y}"
+        )
+
+
+class Concat(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        return x + y
+
+
+class Merge(Merger):
+    def _merge(self, name: str, x: T, y: T) -> T:
+        return merge(x, y)
+
+
+class DictMerge(Merger):
+    def __init__(self, value_merger: Merger = Forbid):
+        self.value_merger = value_merger
+
+    def _merge(self, name: str, x: T, y: T) -> T:
+        result = copy(x)
+        for key, value in y.items():
```

Comment:
I don't know how to make annotations works in case where T may be str or dict. Maybe use exact types.